cvat/apps/engine/rq.py

```diff
@@ -0,0 +1,527 @@
+
+        return {k: v for k, v in self._job.meta.items() if k not in resettable_fields}
+
+
+class RQMetaWithFailureInfo(AbstractRQMeta):
+    formatted_exception = MutableRQMetaAttribute(
+        RQJobMetaField.FORMATTED_EXCEPTION,
+        validator=lambda x: isinstance(x, str),
+        optional=True,
+    )
+    exc_type = MutableRQMetaAttribute(
+        RQJobMetaField.EXCEPTION_TYPE,
+        validator=lambda x: issubclass(x, BaseException),
+        optional=True,
+    )
+    exc_args = MutableRQMetaAttribute(
+        RQJobMetaField.EXCEPTION_ARGS,
+        validator=lambda x: isinstance(x, tuple),
+        optional=True,
+    )
+
+    @staticmethod
+    def _get_resettable_fields() -> list[RQJobMetaField]:
+        return [
+            RQJobMetaField.FORMATTED_EXCEPTION,
+            RQJobMetaField.EXCEPTION_TYPE,
+            RQJobMetaField.EXCEPTION_ARGS,
+        ]
+
+
+class BaseRQMeta(RQMetaWithFailureInfo):
+    # immutable && required fields
+    user: UserMeta = UserRQMetaAttribute()
+    request: RequestMeta = RequestRQMetaAttribute()
+
+    # immutable && optional fields
+    org_id: int | None = ImmutableRQMetaAttribute(
+        RQJobMetaField.ORG_ID, validator=lambda x: isinstance(x, int), optional=True
+    )
+    org_slug: int | None = ImmutableRQMetaAttribute(
+        RQJobMetaField.ORG_SLUG, validator=lambda x: isinstance(x, str), optional=True
+    )
+    project_id: int | None = ImmutableRQMetaAttribute(
+        RQJobMetaField.PROJECT_ID, validator=lambda x: isinstance(x, int), optional=True
+    )
+    task_id: int | None = ImmutableRQMetaAttribute(
+        RQJobMetaField.TASK_ID, validator=lambda x: isinstance(x, int), optional=True
+    )
+    job_id: int | None = ImmutableRQMetaAttribute(
+        RQJobMetaField.JOB_ID, validator=lambda x: isinstance(x, int), optional=True
+    )
+
+    # mutable && optional fields
+    progress: float | None = MutableRQMetaAttribute(
+        RQJobMetaField.PROGRESS, validator=lambda x: isinstance(x, float), optional=True
+    )
+    status: str | None = MutableRQMetaAttribute(
+        RQJobMetaField.STATUS, validator=lambda x: isinstance(x, str), optional=True
+    )
+
+    @staticmethod
+    def _get_resettable_fields() -> list[RQJobMetaField]:
+        return RQMetaWithFailureInfo._get_resettable_fields() + [
+            RQJobMetaField.PROGRESS,
+            RQJobMetaField.STATUS,
+        ]
+
+    @classmethod
+    def build(
+        cls,
+        *,
+        request: ExtendedRequest,
+        db_obj: Model | None,
+    ):
+        # to prevent circular import
+        from cvat.apps.events.handlers import job_id, organization_slug, task_id
+        from cvat.apps.webhooks.signals import organization_id, project_id
+
+        oid = organization_id(db_obj)
+        oslug = organization_slug(db_obj)
+        pid = project_id(db_obj)
+        tid = task_id(db_obj)
+        jid = job_id(db_obj)
+
+        user = request.user
+
+        return cls.for_meta(
```

Comment:
Isn't the class redundant here? You're creating a dictionary, wrapping it, then immediately unwrapping it.

Reply:
Okay, right now there is no reason to do so. In the previous implementation, I thought it would be better to check that the generated meta matches the expected structure.